src/Substance.hs

```diff
@@ -4,535 +4,450 @@
-check :: SubProg -> SubEnv
-check p = let env1 = foldl checkDecls initE p
-              env2 = foldl checkReferencess env1 p
-              in
-            env2 { subObjs = reverse $ subObjs env2 }
-          where initE = SubEnv { subObjs = [], subDefs = M.empty,  subSymbols = M.empty, subApps = [], subArgs = M.empty }
-
-applyDef (n, m) d = case M.lookup n d of
+-- AST as the input and the environment recived from the DSLL type checker, checks the validity of the program acoording to
+-- the typechecking rules, and outputs a collection of information.
+check :: SubProg -> VarEnv -> VarEnv
+check p varEnv = let env = foldl checkSubStmt varEnv p
+                 in if (null (errors env))
+                    then env
+                    else error ("Substance type checking failed with the following problems: \n" ++ (errors env))
+
+checkSubStmt :: VarEnv -> SubStmt -> VarEnv
+checkSubStmt varEnv  (Decl t (VarConst n)) = let env  = checkT varEnv t
+                                                 env1 = addDeclaredName n env
+                                             in env1 { varMap = M.insert (VarConst n) t $ varMap env1 }
+
+checkSubStmt varEnv  (Bind v e) = let (vstr, vt) = checkVarE varEnv v
+                                      (estr, et) = checkExpression varEnv e -- TODO: Check lazy evaluation on et
+                                  in if (isJust vt && isJust et && vt /= et)
+                                     then varEnv { errors = (errors varEnv) ++ vstr ++ estr ++ "Expression of type " 
+                                                   ++ (show et)
+                                                   ++ " assigned to variable of type " ++ (show vt) ++ "\n"}
+                                     else varEnv { errors = (errors varEnv) ++ vstr ++ estr }
+
+checkSubStmt varEnv  (ApplyP p) = checkPredicate varEnv p
+
+checkPredicate :: VarEnv -> Predicate -> VarEnv
+checkPredicate varEnv (Predicate (PredicateConst p) args pos) =
+               case checkAndGet p (predicates varEnv) pos of
+               Right p -> case p of
+                 Pred1 p1 -> checkVarPred varEnv args p1
+                 Pred2 p2 -> checkRecursePred varEnv args
+               Left _ -> case checkAndGet p (operations varEnv) pos of
+                 Right o -> checkVarOperation varEnv args o
+                 Left err -> varEnv { errors = (errors varEnv) ++ err }
+
+checkVarPred :: VarEnv -> [PredArg] -> Predicate1 -> VarEnv
+checkVarPred varEnv args (Predicate1 name yls kls tls _) =
+             let exprArgs      = map isVarPredicate args
+                 errAndTypesLs = map (checkExpression varEnv) exprArgs
+                 errls         = map (\(err1, t1) -> err1) errAndTypesLs
+                 err           = foldl (\err1 err2 -> err1 ++ err2) "" errls
+                 argTypes      = map (\(err1, t1) -> t1) errAndTypesLs
+             in if (foldl (\b at1 -> b && isJust at1) True argTypes)
```

Comment:
what is this condition trying to test? please refactor into a descriptively named function